xds/internal/balancer/balancergroup/balancergroup.go

```diff
@@ -567,10 +591,23 @@ func buildPickerAndState(m map[internal.LocalityID]*pickerState) balancer.State
 	default:
 		aggregatedState = connectivity.TransientFailure
 	}
-	if aggregatedState == connectivity.TransientFailure {
-		return balancer.State{ConnectivityState: aggregatedState, Picker: base.NewErrPicker(balancer.ErrTransientFailure)}
+
+	// Make sure picker's return error is consistent with the aggregatedState.
+	//
+	// TODO: This is true for balancers like weighted_target, but not for
+	// routing. For routing, we want to always build picker with all sub-pickers
+	// (not even ready sub-pickers), so even if the overall state is Ready, pick
+	// for certain RPCs can behave like Connecting or TransientFailure.
+	var picker balancer.Picker
+	switch aggregatedState {
+	case connectivity.TransientFailure:
+		picker = base.NewErrPicker(balancer.ErrTransientFailure)
```

Comment:
`ErrTransientFailure` is now deprecated; can this be updated easily?  (If not we can change it later.)

Reply:
Made it a local const error.

Reply:
This turns out to be not easy, because tests in parent balancers are expecting this error. Change reverted.